modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/services/StateChannelBinarySender.scala

```diff
@@ -0,0 +1,200 @@
+          .warn(
+            s"${pending.length} state channel binaries are still not confirmed on the global state. Waiting for the next global snapshot."
+          )
+          .whenA(pending.nonEmpty)
+      } yield ()
+
+      def processPending: F[Unit] = {
+        val retry = for {
+          pending <- pendingR.get
+          avgConfirmedPerSnapshot <- avgConfirmedCountPerSnapshotR.get
+          cap = Math
+            .round(avgConfirmedPerSnapshot.toDouble + (avgConfirmedPerSnapshot.toDouble / 2))
+            .toInt // Try sending 50% more than avg confirmed
+          toRetry = pending.take(cap)
+          _ <- toRetry.traverse(tracked => post(tracked.binary))
+        } yield ()
+
+        catchUpMode.get.ifM(retry, Applicative[F].unit)
+      }
+
+      def clearPending: F[Unit] = pendingR.set(Queue.empty)
+
+      private def updateExpectedOrdinalsToConfirm(
+        confirmedTracked: Queue[TrackedBinary],
+        snapshotOrdinal: SnapshotOrdinal
+      ): F[PosLong] =
+        expectedOrdinalsToConfirmR.updateAndGet { prev =>
+          if (confirmedTracked.isEmpty) prev
+          else {
+            val ordinalsPassed = confirmedTracked.map(tb => snapshotOrdinal.value - tb.sentAtOrdinal.value)
+            val avgOrdinalsPassed = PosLong.from(Math.round(ordinalsPassed.sum.toDouble / ordinalsPassed.length)).getOrElse(prev)
+            avgOrdinalsPassed
+          }
+        }
+
+      private def updateAvgConfirmedCountPerSnapshot(confirmedTracked: Queue[TrackedBinary]): F[Unit] =
+        avgConfirmedCountPerSnapshotR.update { prev =>
+          NonNegLong.from {
+            Math.ceil((prev.toDouble + confirmedTracked.length.toDouble) / 2.0).toLong
+          }.getOrElse(prev)
+        }
+
+      private def updateCatchUpMode(
+        pending: Queue[TrackedBinary],
+        snapshotOrdinal: SnapshotOrdinal,
+        expectedOrdinalsToConfirm: PosLong
+      ): F[Unit] =
+        catchUpMode.update(prev =>
+          if (pending.isEmpty) {
+            false
+          } else {
+            val hasStalled = pending.exists { p =>
+              val ordinalsWithoutConfirmationCount =
+                NonNegLong
+                  .from(snapshotOrdinal.value - p.sentAtOrdinal.value)
+                  .getOrElse(NonNegLong.MaxValue)
+              ordinalsWithoutConfirmationCount >= expectedOrdinalsToConfirm
+            }
+            hasStalled || prev
+          }
+        )
+
+      private def enqueue(binary: Hashed[StateChannelSnapshotBinary]): F[Unit] =
+        for {
+          currentOrdinal <- lastGlobalSnapshotStorage.getOrdinal.map(_.getOrElse(SnapshotOrdinal.MinValue))
+          _ <- pendingR.update { pending =>
+            pending.enqueue(TrackedBinary(binary, currentOrdinal))
+          }
+        } yield ()
+
+      private def post(binary: Hashed[StateChannelSnapshotBinary]): F[Unit] = {
+        val sendRetries = 5
+
+        val retryPolicy: RetryPolicy[F] = RetryPolicies.limitRetries(sendRetries)
```

Comment:
This looks like it retries immediately.
Are the causes of failure such that we can do that?

Reply:
There is no need to delay because we use `globalL0ClusterStorage.getRandomPeer` inside. So if it fails then we immediately retry to another peer selected randomly.